Add a draft of a user guide on writing custom extractors (#282)
datalad/datalad-metalad: docs/source/user_guide/writing-extractors.rst

.. -*- mode: rst -*-
.. vi: set ft=rst sts=4 ts=4 sw=4 et tw=79:

.. _chap_writing-extractors:

Writing custom extractors
*************************

An extractor, in MetaLad sense, is a class derived from one of the base extractor classes defined in ``datalad_metalad.extractors.base`` (``DatasetMetadataExtractor`` or ``FileMetadataExtractor``).
It needs to implement several required methods, most notably ``extract()``.

Example extractors can be found in MetaLad source code.

.. TODO:: Add links to the example extractors

Base class
==========

There are two primary types of extractors, dataset-level and file-level.

Dataset-level extractors, by inheritance from the ``DatasetMetadataExtractor`` class, can access the dataset on which they operate as ``self.dataset``. Extractor functions may use it to call any dataset methods.

> File-level extractors, by inheritance from the ``FileMetadataExtractor``, can refer to the file on which it operates through ``self.file_info``.

Author: In the section above, I wanted to clearly state the primary ways to query the dataset content. While dataset methods are fairly clear to me, I have pretty much no experience on `file_info` - maybe someone could add some details on what it can do (or a link to api docs).

Author: Added a little bit in 01622f8.

A UUID can be generated in several ways, e.g. with Python (``import uuid; print(uuid.uuid4())``), or online generators
(e.g. https://www.uuidgenerator.net/ or by searching 'UUID' in `DuckDuckGo <https://duckduckgo.com/>`_).

While version 4 (random) UUIDs are probably sufficient, version 5 UUIDs (generated based on a given namespace and name) can also be used.
For example, extractors in packages developed under the datalad organisation could use ``<extractor_name>.datalad.org``.
Example generation with Python: ``uuid.uuid5(uuid.NAMESPACE_DNS, "bids_dataset.datalad.org")``.

Example::

  def get_id(self) -> UUID:
      return UUID("d1e3f728-7868-4f33-a4a5-0f9f57dce342")

``get_version()``
-----------------

This function should return a version string representing the extractor's version.

The extractor version is meant to be included with the extracted metadata.
It might be a version number, or a reference to a particular schema.
While there are no requirements on the versioning scheme, it should be chosen carefully, and rules for version updating need to be considered.

> For example, when using `semantic versioning <https://semver.org/>`_ style, one might decide to guarantee that attribute type and name will never change within a major release, while new attributes may be added with a minor release.

Author: This ruleset borrowed from https://pypi.org/project/nbib/ which I used recently.

Its purpose is to make sure that content required for metadata extraction is present
(relevant, for example, if some of files to be inspected may be annexed).
The function should return ``True`` if it has obtained the required content, or confirmed its presence.
If it returns ``False``, metadata extraction will not proceed.

This function can be a place to call ``dataset.get()``.
It is advisable to disable result rendering (``result_renderer="disabled"``), because during metadata extraction, users will typically want to redirect standard output to a file or another command.

Example::

  def get_required_content(self) -> bool:
      result = self.dataset.get("CITATION.cff", result_renderer="disabled")
      return result[0]["status"] in ("ok", "notneeded")

``is_content_required()``
-------------------------

This function is used in file-extractors only.
It is a file-extractor counterpart to ``get_required_content``.
Its purpose is to tell MetaLad if the file content is required or not
(relevant for annexed files - extraction may depend on file content, or require only annex key).
If the function returns ``True``, MetaLad will get the file content.

> If it returns ``False``, the get operation will not be performed.

Author: I didn't work with file-level extractors, that's my guess - is that true?

Also, might be good to come up with an example.

Author: Confirmed, this is how it works.

(relevant, for example, if some of files to be inspected may be annexed).
The function should return ``True`` if it has obtained the required content, or confirmed its presence.
If it returns ``False``, metadata extraction will not proceed.

This function can be a place to call ``dataset.get()``.
It is advisable to disable result rendering (``result_renderer="disabled"``), because during metadata extraction, users will typically want to redirect standard output to a file or another command.

Example::

  def get_required_content(self) -> bool:
      result = self.dataset.get("CITATION.cff", result_renderer="disabled")
      return result[0]["status"] in ("ok", "notneeded")

``is_content_required()``
-------------------------

This function is used in file-extractors only.
It is a file-extractor counterpart to ``get_required_content()``.
Its purpose is to tell MetaLad whether the file content is required or not
(relevant for annexed files - extraction may depend on file content, or require only annex key).
If the function returns ``True``, MetaLad will get the file content.
If it returns ``False``, the get operation will not be performed.
      
``extract()``
-------------

> This function is used for actual metadata extraction, and should return an ``datalad_metalad.extractors.base.ExtractorResult`` object.

Reviewer: ```suggestion
This function is used for actual metadata extraction. It has one parameter called ``output_location``. If the output category of the extractor is ``DataOutputCategory.IMMEDIATE``, this parameter will be ``None``. If the output category is ``DataOutputCategory.FILE``. this parameter will contain either a file name or a ``file``-object into which the extractor can write its output.

 The function should return an ``datalad_metalad.extractors.base.ExtractorResult`` object.
```

Author: Okay, I think I get it. So the extractor declares its output category to MetaLad... and then, if it declared FILE, MetaLad can use this parameter to tell the extractor what file it should write to.

.. -*- mode: rst -*-
.. vi: set ft=rst sts=4 ts=4 sw=4 et tw=79:

.. _chap_writing-extractors:

Writing custom extractors
*************************

> Metalad supports the automated extraction of metadata through a common single interface that allows the execution of metadata extractors.

Author: ```suggestion
Metalad supports automated extraction of metadata through a common single interface which allows execution of metadata extractors.
```

(relevant, for example, if some of files to be inspected may be annexed).
The function should return ``True`` if it has obtained the required content, or confirmed its presence.
If it returns ``False``, metadata extraction will not proceed.

This function can be a place to call ``dataset.get()``.
It is advisable to disable result rendering (``result_renderer="disabled"``), because during metadata extraction, users will typically want to redirect standard output to a file or another command.

Example::

  def get_required_content(self) -> bool:
      result = self.dataset.get("CITATION.cff", result_renderer="disabled")
      return result[0]["status"] in ("ok", "notneeded")

``is_content_required()``
-------------------------

This function is used in file-extractors only.
It is a file-extractor counterpart to ``get_required_content()``.
Its purpose is to tell MetaLad whether the file content is required or not
(relevant for annexed files - extraction may depend on file content, or require only annex key).
If the function returns ``True``, MetaLad will get the file content.
If it returns ``False``, the get operation will not be performed.
      
``extract()``
-------------

> This function is used for actual metadata extraction. It has one parameter called ``output_location``. If the output category of the extractor is ``DataOutputCategory.IMMEDIATE``, this parameter will be ``None``. If the output category is ``DataOutputCategory.FILE``. this parameter will contain either a file name or a ``file``-object into which the extractor can write its output.

Author: ```suggestion
This function is used for actual metadata extraction. It has one parameter called ``output_location``. If the output category of the extractor is ``DataOutputCategory.IMMEDIATE``, this parameter will be ``None``. If the output category is ``DataOutputCategory.FILE``, this parameter will contain either a file name or a ``file``-object into which the extractor can write its output.
```

>  The function should return an ``datalad_metalad.extractors.base.ExtractorResult`` object.

Author: ```suggestion
The function should return an ``datalad_metalad.extractors.base.ExtractorResult`` object.
```

Looks like an "unexpected indent" made its way through and was caught by the CI.
